dali/test/python/operator_1/test_clahe.py

```diff
@@ -0,0 +1,586 @@
+            # RGB image - apply to luminance channel only
+            # Use LAB color space to match DALI exactly
+            lab = cv2.cvtColor(image, cv2.COLOR_RGB2Lab)
+            # Apply CLAHE to L channel
+            lab[:, :, 0] = clahe.apply(lab[:, :, 0])
+            # Convert back to RGB
+            result = cv2.cvtColor(lab, cv2.COLOR_Lab2RGB)
+        elif image.shape[2] == 3:
+            # Apply CLAHE to each RGB channel separately
+            result = np.zeros_like(image)
+            for i in range(3):
+                result[:, :, i] = clahe.apply(image[:, :, i])
+        else:
+            raise ValueError(f"Unsupported image shape: {image.shape}")
+    else:
+        # Grayscale
+        result = clahe.apply(image)
+
+    return result
+
+
+class MemoryPipeline(Pipeline):
+    """DALI pipeline using external data input for exact comparison"""
+
+    def __init__(self, image_array, tiles_x=8, tiles_y=8, clip_limit=2.0, device="gpu"):
+        super().__init__(batch_size=1, num_threads=1, device_id=0)
+        self.image_array = image_array
+        self.tiles_x = tiles_x
+        self.tiles_y = tiles_y
+        self.clip_limit = clip_limit
+        self.device = device
+
+    def define_graph(self):
+        # Use external source to feed exact same data as OpenCV
+        images = fn.external_source(
+            source=lambda: [self.image_array],
+            device="cpu",
+            ndim=len(self.image_array.shape),
+        )
+
+        if self.device == "gpu":
+            # Move to GPU for processing
+            images_processed = images.gpu()
+        else:
+            # Keep on CPU for processing
+            images_processed = images
+
+        # Apply CLAHE operator
+        clahe_result = fn.clahe(
+            images_processed,
+            tiles_x=self.tiles_x,
+            tiles_y=self.tiles_y,
+            clip_limit=float(self.clip_limit),
+            luma_only=True,
+            device=self.device,
+        )
+
+        return clahe_result
+
+
+def apply_dali_clahe_from_memory(image_array, tiles_x=8, tiles_y=8, clip_limit=2.0, device="gpu"):
+    """Apply DALI CLAHE using memory-based pipeline for exact input matching"""
+    pipe = None
+    try:
+        # Create memory-based pipeline
+        pipe = MemoryPipeline(image_array, tiles_x, tiles_y, clip_limit, device)
+        pipe.build()
+
+        # Run pipeline
+        outputs = pipe.run()
+        result = outputs[0].as_cpu().as_array()[0]  # Get first image from batch
+
+        # Enhanced data type conversion with rounding for better precision
+        if result.dtype != np.uint8:
+            # Round to nearest integer before clipping for better accuracy
+            result = np.round(np.clip(result, 0, 255)).astype(np.uint8)
+
+        return result
+
+    finally:
+        # Explicitly release pipeline resources
+        if pipe is not None:
+            del pipe
+
+
+class ClahePipeline(Pipeline):
```

Comment:
Wouldn't it be easier to use `@pipeline_def` decorator instead of inherting from `Pipeline`?